File: app/src/main/java/com/github/oryanmat/trellowidget/util/DateTimeUtil.java
package com.github.oryanmat.trellowidget.util;

import android.support.annotation.NonNull;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Locale;
import java.util.TimeZone;

public class DateTimeUtil {
    static final String DATE_PARSE_ERROR = "Bad date";

    static final SimpleDateFormat apiFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.getDefault());
    static final SimpleDateFormat widgetFormat = new SimpleDateFormat("MMM dd", Locale.getDefault());
    static final SimpleDateFormat activityFormat = new SimpleDateFormat("MMM dd 'at' HH:mm a", Locale.getDefault());

    static {
        // it appears that trello returns dates in UTC time zone
        apiFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
    }

    public static String parseDate(@NonNull String date) {
        return parse(date, widgetFormat);
    }

    public static String parseDateTime(@NonNull String date) {
        return parse(date, activityFormat);
    }

    private static String parse(@NonNull String date, DateFormat format) {
        try {
            return format.format(apiFormat.parse(date));
        } catch (ParseException e) {
            return DATE_PARSE_ERROR;
        }
    }
}

File: app/src/main/kotlin/com/github/oryanmat/trellowidget/util/DateTimeUtil.kt
package com.github.oryanmat.trellowidget.util

import java.text.DateFormat
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*

object DateTimeUtil {
    internal val DATE_PARSE_ERROR = "Bad date"

    internal val apiFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.getDefault())
    internal val widgetFormat = SimpleDateFormat("MMM dd", Locale.getDefault())
    internal val activityFormat = SimpleDateFormat("MMM dd 'at' HH:mm a", Locale.getDefault())

    init {
        // it appears that trello returns dates in UTC time zone
        apiFormat.timeZone = TimeZone.getTimeZone("UTC")
    }

    fun parseDate(date: String) = parse(date, widgetFormat)

    fun parseDateTime(date: String) = parse(date, activityFormat)

    private fun parse(date: String, format: DateFormat) : String = try {
        format.format(apiFormat.parse(date))
    } catch (e: ParseException) {
        DATE_PARSE_ERROR
    }
}